ov/server_hardware.go

```diff
@@ -359,3 +479,231 @@ func (c *OVClient) GetServerFirmwareByUri(uri utils.Nstring) (ServerFirmware, er
 }
+
+// GetServerFirmwareByUri gets firmware for a server hardware with uri
+func (c *OVClient) GetServerFirmwareList(filters []string, sort string, start string, count string) (ServerFirmwareList, error) {
+	var (
+		uri          = "/rest/server-hardware/*/firmware"
+		q            map[string]interface{}
+		firmwareList ServerFirmwareList
+	)
+
+	q = c.SetQueryParams(filters, sort, start, count, "")
+
+	// refresh login
+	c.RefreshLogin()
+	c.SetAuthHeaderOptions(c.GetAuthHeaderMap())
+
+	data, err := c.RestAPICall(rest.GET, uri, nil, q)
+	if err != nil {
+		return firmwareList, err
+	}
+
+	log.Debugf("GetServerFirmwareList %s", data)
+
+	if err := json.Unmarshal([]byte(data), &firmwareList); err != nil {
+		return firmwareList, err
+	}
+
+	if count == "" {
+		total := strconv.Itoa(firmwareList.Total)
+		return c.GetServerFirmwareList(filters, "", "", total)
+	}
+
+	for i := 0; i < firmwareList.Count; i++ {
+		firmwareList.Members[i].Client = c
+	}
+
+	return firmwareList, nil
+}
+
+// Refresh server hardware configuration
+func (c *OVClient) RefreshServerHardware(id string, hardware ServerHardware) error {
+	var (
+		uri = "/rest/server-hardware/" + id + "/refreshState"
+		t   *Task
+	)
+	// refresh login
+	c.RefreshLogin()
+	c.SetAuthHeaderOptions(c.GetAuthHeaderMap())
+
+	log.Debugf("REST : %s \n %+v\n", uri, hardware)
+	data, err := c.RestAPICall(rest.PUT, uri, hardware)
+	if err != nil {
+		log.Errorf("Error while refreshing server hardware: %s", err)
+		return err
+	}
+
+	log.Debugf("Response of server hardware refresh %s", data)
+	if err := json.Unmarshal([]byte(data), &t); err != nil {
+		log.Errorf("Error with task un-marshal: %s", err)
+		return err
+	}
+
+	return nil
+}
+
+// Updates iLO Firmware Version to minimum firmware version
+// supported by Oneview appliance
+func (c *OVClient) UpdateiLOFirmwareVersion(id string) error {
+	var (
+		uri = "/rest/server-hardware/" + id + "/mpFirmwareVersion"
+		t   *Task
+	)
+	// refresh login
+	c.RefreshLogin()
+	c.SetAuthHeaderOptions(c.GetAuthHeaderMap())
+
+	log.Debugf("REST : %s \n %+v\n", uri, nil)
+	data, err := c.RestAPICall(rest.PUT, uri, nil)
+	if err != nil {
+		log.Errorf("Error while updating firmware version: %s", err)
+		return err
+	}
+
+	log.Debugf("Response of firmware version update %s", data)
+	if err := json.Unmarshal([]byte(data), &t); err != nil {
+		log.Errorf("Error with task un-marshal: %s", err)
+		return err
+	}
+
+	return nil
+}
+
+func (c *OVClient) SetOneTimeBoot(serverHardwareId string, value string) error {
+	patchOperation := PatchData{
+		Op:    "replace",
+		Path:  "/oneTimeBoot",
+		Value: value,
+	}
+	operation := []PatchData{patchOperation}
+	fmt.Println("Update server Hardware's oneTimeBoot\n")
+	err := c.Patch(serverHardwareId, operation)
+	if err != nil {
+		log.Errorf("Error while updating oneTimeBoot: %s", err)
+		return err
+	}
+
+	return nil
+}
+
+func (c *OVClient) PatchPowerState(id string, operation []PatchPowerData) error {
```

Comment:
I don't understand why do we need this module when we already have 'Patch' module

Reply:
For /powerState, we need to pass value as map whereas for other operations its a string. So, I have defined separate struct and function to take care of this issue.